ACPs/30-avalanche-warp-x-evm.md

```diff
@@ -0,0 +1,216 @@
+
+### Precompile Implementation
+
+All types, events, and function arguments/outputs are encoded using the ABI package according to the official [Solidity ABI Specification](https://docs.soliditylang.org/en/latest/abi-spec.html).
+
+Like any contract, the precompile can be called by the available CALL opcodes: `CALL`, `DELEGATECALL`, `STATICCALL`, and `CALLCODE`. Note: `DELEGATECALL` is treated as a normal call when calling precompiles.
+
+When the precompile is invoked with a given `calldata` argument, the first four bytes (`calldata[0:4]`) are read as the [function selector](https://docs.soliditylang.org/en/latest/abi-spec.html#function-selector). If the function selector matches the function selector of one of the functions defined by the Solidity interface, the contract invokes the corresponding execution function with the remaining calldata ie. `calldata[4:]`.
+
+For the full specification of the execution functions defined in the Solidity interface, see the reference implementation here:
+
+- [sendWarpMessage](https://github.com/ava-labs/subnet-evm/blob/v0.5.9/x/warp/contract.go#L226)
+- [getVerifiedWarpMessage](https://github.com/ava-labs/subnet-evm/blob/v0.5.9/x/warp/contract.go#L187)
+- [getVerifiedWarpBlockHash](https://github.com/ava-labs/subnet-evm/blob/v0.5.9/x/warp/contract.go#L145)
+- [getBlockchainID](https://github.com/ava-labs/subnet-evm/blob/v0.5.9/x/warp/contract.go#L96)
+
+### Gas Costs
+
+The Warp Precompile charges gas during the verification of included Avalanche Warp Messages, which is included in the intrinsic gas cost of the transaction, and during the execution of the precompile.
+
+#### Verification Gas Costs
+
+Pre-verification charges the following costs for each Avalanche Warp Message:
+
+- GasCostPerSignatureVerification: 20000
+- GasCostPerWarpMessageBytes: 100
+- GasCostPerWarpSigner: 500
+
+These numbers were determined experimentally using the benchmarks available [here](https://github.com/ava-labs/subnet-evm/blob/master/x/warp/predicate_test.go#L687) to target approximately the same mgas/s as existing precompile benchmarks in the EVM, which ranges between 50-200 mgas/s.
+
+In addition to the benchmarks, the following assumptions and goals were taken into account:
+
+- BLS Public Key Aggregation is extremely fast, resulting in charging more for the base cost of a single BLS Multi-Signature Verification than each individual key
+- The cost per byte included in the transaction should be strictly higher for including Avalanche Warp Messages than via transaction calldata, so that the Warp Precompile does not change the worst case maximum block size
+
+Note: the verification cost of performing a validator set lookup on the P-Chain is currently excluded. The cost of this lookup is variable depending on how old the referenced P-Chain height is from the perspective of each validator.
```

Comment:
The proper way to charge or not charge for P-Chain validator set lookups is the biggest question imo.

If this is fully parallelized, it may not increase block verification latency at all (really nice property), but reasoning about the worst case is difficult here.

Reply:
I think this should be included in `Open Questions` instead of what is there. What is there could be moved to a FAQ section under this section maybe?